samza-operator/src/main/java/org/apache/samza/operators/impl/OperatorImpls.java

```diff
@@ -0,0 +1,123 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.operators.impl;
+
+
+import org.apache.samza.operators.MessageStream;
+import org.apache.samza.operators.MessageStreamImpl;
+import org.apache.samza.operators.data.Message;
+import org.apache.samza.operators.spec.OperatorSpec;
+import org.apache.samza.operators.spec.PartialJoinOperatorSpec;
+import org.apache.samza.operators.spec.SinkOperatorSpec;
+import org.apache.samza.operators.spec.StreamOperatorSpec;
+import org.apache.samza.operators.spec.WindowOperatorSpec;
+import org.apache.samza.operators.windows.WindowOutput;
+import org.apache.samza.operators.windows.WindowState;
+import org.apache.samza.task.TaskContext;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+
+/**
+ * Instantiates the DAG of {@link OperatorImpl} implementations corresponding to the {@link OperatorSpec}s.
+ */
+public class OperatorImpls {
+
+  /**
+   * Holds the mapping between the {@link OperatorSpec} and {@link OperatorImpl}s instances.
+   */
+  private static final Map<OperatorSpec, OperatorImpl> OPERATOR_IMPLS = new ConcurrentHashMap<>();
+
+  /**
+   * Traverses the DAG of {@link OperatorSpec} starting from the provided {@link MessageStreamImpl}, creates the
+   * corresponding DAG of {@link OperatorImpl}s, and returns its root {@link NoOpOperatorImpl} node.
+   *
+   * @param source  the input {@link MessageStreamImpl} to instantiate {@link OperatorImpl}s for
+   * @param <M>  type of the {@link Message} in the input {@link MessageStreamImpl}
+   * @param context  the {@link TaskContext} required to instantiate operators
+   * @return  root node for the {@link OperatorImpl} DAG.
+   */
+  public static <M extends Message> NoOpOperatorImpl<M> createOperatorImpls(MessageStreamImpl<M> source, TaskContext context) {
+    // since the source message stream might have multiple operator specs registered on it,
+    // create a new root node as a single point of entry for the DAG.
+    NoOpOperatorImpl<M> rootOperator = new NoOpOperatorImpl<>();
+    // create the pipeline/topology starting from the source
+    source.getRegisteredOperatorSpecs().forEach(registeredOperator -> {
+        // pass in the source and context s.t. stateful stream operators can initialize their stores
+        OperatorImpl<M, ? extends Message> operatorImpl =
+            createAndRegisterOperatorImpl(registeredOperator, source, context);
+        rootOperator.registerNextOperator(operatorImpl);
+      });
+    return rootOperator;
+  }
+
+  /**
+   * Helper method to traverse the {@link OperatorSpec} DAG and instantiate and link the corresponding {@link OperatorImpl}s.
```

Comment:
You're right. I don't feel strongly about it in this case.

As a personal preference, I add implementation notes to methods that are particularly meaty. In those cases, I usually add a Notes section in the method. (javadoc or otherwise) For example,  http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b14/java/util/Arrays.java#1093